### DISCUSSION TITLE
Sending failed when the amount of sequence type data is large

### DISCUSSION THREAD

**Discussion by zip-link** [eProsima/Fast-DDS#2567] [Q&A]
Using the default QoS configuration of version 2.5.0, when transmitting sequence data, when the number of elements of sequence is relatively large, it will fail to send after matching the subscriber. After releasing printing, serialization failed. How to solve this problem?

There are warnings below：
2022-03-10 10:23:54.115 [RTPS_WRITER Info] 0.0.3.c2 Sending Heartbeat (1 - 1) -> Function send_heartbeat_nts_
2022-03-10 10:23:54.929 [RTPS_WRITER Warning] RTPSWriter:Serialization returns false -> Function perform_create_new_change
2022-03-10 10:23:56.917 [RTPS_WRITER Info] Sending relevant changes as DATA/DATA_FRAG messages -> Function add_data
2022-03-10 10:23:56.917 [RTPS_WRITER Info] Sending INFO_TS message -> Function add_info_ts_in_buffer

**Comment by EduPonz** [comment]
Hi @zip-link ,

We would need more information to reproduce your issue:

* Which OS are you using?
* How are you compiling Fast DDS (CMake flags, etc.)?
* What is a relative large number of elements for a sequence?

Ideally, some short of reproducer would be nice; we normally ask for a Dockerfile, as that way the reproducer is self contained.

Thanks!

**Reply by zip-link** [reply]
First answer your question
1. Using Ubuntu system
2. Cmake has now released statistics and debug
3. The amount of data with this problem is currently vague

for example
Here is my IDL definition，

```idl
struct struA
{
    unsigned long index;
    string message;
};
struct seqA
{
    sequence<struA> seq1;
};
```
When the content is more than 2000, the problem will be repeated
```c++
seqA a;
for( i= 0; i < 2000; i++)
{
    a.seq1().push_back(strua);
}
``` 
But strangely, when the IDL definition becomes the following, the same assignment program and release program, the problem will not recur
```idl
struct seqA
{
    sequence<struA> seq1;
    sequence<struB> seq2;
};
```
In fact, seq2 is not used, so I suspect that there is a boundary problem in the serialization code of variable length data types, but it is also strange that when I increase the length of seq1 to 4000, the problem is repeated by using the second IDL definition

**Reply by EduPonz** [reply]
Hi @zip-link ,

First, I've taken the liberty of editing your answer to format the code, hope that's OK. In the second IDL, I get that the second sequence is also of `struA`, right? Would you mind updating it?

I do have one more question, which version of Fast DDS-Gen are you using to generate the type support?

**Reply by zip-link** [reply]
Yes, the second sequence is actually independent of the type. As long as there is a second sequence, it will be different from the original result. The fastddsgen version uses version 2.0.0. Now I also try to update the gen version

**Reply by LakeFusion** [reply]
I've also have this issue. It seems if there is no maximum size definition for a sequence, fastddsgen simply takes 100 as the limit.

To override the limit of 100 elements, you can define a constant and use it as maximum size for the sequence like this:

```
const int32 MAX_SIZE = 2097152;
...
sequence<uint8, MAX_SIZE> data;
```

**Reply by zip-link** [reply]
Hi,@EduPonz 
I tried to use the latest fastddsgen version 2.1.1 to generate serialization code, but this problem still exists. I think there is still a problem of boundary recognition in the code, but I won't understand it for a while

**Reply by zip-link** [reply]
@LakeFusion thank you,
I've tried your method. It's really useful. You can specify the size when you don't know the IDL definition before. Is this the standard IDL definition method?

**Reply by LakeFusion** [reply]
No problem @zip-link . I don't think it is the standard way but it is a valid IDL syntax which worked for my case. Somehow ROS2 IDLs just work without such limits.

**Reply by MiguelCompany** [reply]
@LakeFusion @zip-link That is standard IDL syntax. On IDL 4.2, the grammar for sequence is defined as

```
<sequence_type> ::= "sequence" "<" <type_spec> "," <positive_int_const> ">"
                |   "sequence" "<" <type_spec> ">"
```
